File: manifold_sampling/m/general_nonsmooth_h_funs/TestCreateCensoredL1LossHfun.m
% From MATLAB and the directory containing this file, execute
%     >> runtests
%
% If you would like to run this from a different folder that includes these
% tests as a subfolder, then from that folder execute
%     >> runtests("IncludeSubfolders", true)
% Other tests run during execution might require adding other folders to the
% path.
%
% To execute the test suite with coverage enabled and to generate an HTML-format
% coverage report, execute from /path/to/IBCDFO/pounders/m
%     >> runtests("IncludeSubfolders", true, "ReportCoverageFor", pwd)
%

classdef TestCreateCensoredL1LossHfun < matlab.unittest.TestCase
    properties
        C
        D
        hfun
        Z
        hF
        grads
        Hash
        C_short
        D_short
        path_orig
    end

    methods (TestMethodSetup)

        function setup(testCase)
            % tests need product_of_cells
            [here_path, ~, ~] = fileparts(mfilename('fullpath'));
            testCase.path_orig = addpath(fullfile(here_path, '..'));

            % Confirm that we have good arguments for use in testing
            M = 10;

            testCase.C = 1:M;
            testCase.D = -1.1 * (1:M);
            [testCase.hfun] = create_censored_L1_loss_hfun(testCase.C, ...
                                                           testCase.D);

            % The test values were not determined by hand and are, therefore,
            % not known to be correct.  Rather, they were gathered from test
            % output at some time and are used here to catch regressions and to
            % confirm that the Python and MATLAB code are returning the same
            % values for the same problems.
            testCase.Z = 2.1 * (0:length(testCase.C) - 1);
            [testCase.hF, testCase.grads, testCase.Hash] = ...
                testCase.hfun(testCase.Z);
            testCase.assertEqual(156.0, testCase.hF);
            expected = ones([M 1]);
            expected(1) = 0;
            testCase.assertEqual(testCase.grads, expected);
            testCase.assertTrue(iscell(testCase.Hash));
            [tmp, n_hash] = size(testCase.Hash);
            testCase.assertEqual(tmp, 1);
            testCase.assertEqual(n_hash, 1);
            testCase.assertEqual(size(testCase.grads), [M, 1]);
            expected = {'2111111111'};
            testCase.assertEqual(testCase.Hash, expected);

            % Sanity check hash result/H0
            [hF_H0, grads_H0] = testCase.hfun(testCase.Z, testCase.Hash);
            testCase.assertEqual(hF_H0, testCase.hF);
            testCase.assertEqual(grads_H0, testCase.grads);

            % Good but different size for testing size incompatibilities
            testCase.C_short = [1.1 2.2 -3.3];
            testCase.D_short = [3.3 -1.1 -2.2];
            M_short = length(testCase.C_short);
            testCase.assertNotEqual(length(testCase.C_short), length(testCase.C));
            testCase.assertNotEqual(length(testCase.D_short), length(testCase.D));

            [hfun] = create_censored_L1_loss_hfun(testCase.C_short, ...
                                                  testCase.D_short);

            % As above in terms of correctness and use of expected results
            % here.
            Z_short = 2.1 * (1:length(testCase.C_short));
            [hF, grads, Hash] = hfun(Z_short);
            testCase.assertEqual(15.0, hF);
            expected = ones([M_short 1]);
            expected(1) = -1;
            testCase.assertEqual(grads, expected);
            testCase.assertTrue(iscell(Hash));
            [tmp, n_hash] = size(Hash);
            testCase.assertEqual(tmp, 1);
            testCase.assertEqual(n_hash, 1);
            testCase.assertEqual(size(grads), [M_short, 1]);
            expected = {'311'};
            testCase.assertEqual(Hash, expected);

            % Sanity check hash result/H0
            [hF_H0, grads_H0] = hfun(Z_short, Hash);
            testCase.assertEqual(hF_H0, hF);
            testCase.assertEqual(grads_H0, grads);
        end

    end

    methods (TestMethodTeardown)

        function restorePath(testCase)
            path(testCase.path_orig);
        end

    end

    methods (Test)

        function badArguments(testCase)
            SHAPE_2D = [5 2];

            % Accepts row and column vectors ...
            testCase.verifyTrue(isrow(testCase.C));
            testCase.verifyTrue(isrow(testCase.D));

            C = testCase.C(:);
            testCase.verifyTrue(iscolumn(C));
            [hfun_c] = create_censored_L1_loss_hfun(C, testCase.D);
            [hF_c, grads_c, Hash_c] = hfun_c(testCase.Z);
            testCase.assertEqual(hF_c, testCase.hF);
            testCase.assertEqual(grads_c, testCase.grads);
            testCase.assertEqual(Hash_c, testCase.Hash);

            D = testCase.D(:);
            testCase.verifyTrue(iscolumn(D));
            [hfun_c] = create_censored_L1_loss_hfun(C, D);
            [hF_c, grads_c, Hash_c] = hfun_c(testCase.Z);
            testCase.assertEqual(hF_c, testCase.hF);
            testCase.assertEqual(grads_c, testCase.grads);
            testCase.assertEqual(Hash_c, testCase.Hash);

            % but not matrices
            C_2D = reshape(testCase.C, SHAPE_2D);
            D_2D = reshape(testCase.D, SHAPE_2D);
            testCase.assertEqual(numel(C_2D), length(testCase.C));
            testCase.verifyError( ...
                @()create_censored_L1_loss_hfun(C_2D, testCase.D), ...
                'MATLAB:validation:IncompatibleSize' ...
            );
            testCase.verifyError( ...
                @()create_censored_L1_loss_hfun(testCase.C, D_2D), ...
                'MATLAB:validation:IncompatibleSize' ...
            );
            testCase.verifyError( ...
                @()create_censored_L1_loss_hfun(C_2D, D_2D), ...
                'MATLAB:validation:IncompatibleSize' ...
            );

            % or an empty array
            testCase.verifyError( ...
                @()create_censored_L1_loss_hfun([], testCase.D), ...
                'MATLAB:validators:mustBeNonempty' ...
            );
            testCase.verifyError( ...
                @()create_censored_L1_loss_hfun(testCase.C, []), ...
                'MATLAB:validators:mustBeNonempty' ...
            );
            testCase.verifyError( ...
                @()create_censored_L1_loss_hfun([], []), ...
                'MATLAB:validators:mustBeNonempty' ...
            );

            % C & D must have at least two elements
            testCase.verifyError( ...
                @()create_censored_L1_loss_hfun([1.1], testCase.D), ...
                'POptUS:IncompatibleSizes' ...
            );
            testCase.verifyError( ...
                @()create_censored_L1_loss_hfun(testCase.C, [2.2]), ...
                'POptUS:IncompatibleSizes' ...
            );
            testCase.verifyError( ...
                @()create_censored_L1_loss_hfun([1.1], [2.2]), ...
                'POptUS:ArrayTooShort' ...
            );

            % Finite real values only please
            for bad = ["bad", 1j, 1.0 - 2.0 * 1j]
                testCase.verifyError( ...
                    @()create_censored_L1_loss_hfun(bad, testCase.D), ...
                    'MATLAB:validators:mustBeReal' ...
                );
                testCase.verifyError( ...
                    @()create_censored_L1_loss_hfun(testCase.C, bad), ...
                    'MATLAB:validators:mustBeReal' ...
                );
            end

            for bad = [inf -inf nan]
                testCase.verifyError( ...
                    @()create_censored_L1_loss_hfun(bad, testCase.D), ...
                    'MATLAB:validators:mustBeFinite' ...
                );
                testCase.verifyError( ...
                    @()create_censored_L1_loss_hfun(testCase.C, bad), ...
                    'MATLAB:validators:mustBeFinite' ...
                );
            end

            % C & D must be the same length ...
            testCase.verifyError( ...
                @()create_censored_L1_loss_hfun(testCase.C_short, testCase.D), ...
                'POptUS:IncompatibleSizes' ...
            );
            testCase.verifyError( ...
                @()create_censored_L1_loss_hfun(testCase.C, testCase.D_short), ...
                'POptUS:IncompatibleSizes' ...
            );
        end

        function badZArgument(testCase)
            M = length(testCase.Z);

            % Must be 1D
            for bad = ["bad", 1.1, 1j, []]
                testCase.verifyError( ...
                    @()testCase.hfun(bad), 'POptUS:IncompatibleSizes' ...
                );
            end

            Z_2D = zeros([M, 2]);
            testCase.verifyError( ...
                @()testCase.hfun(Z_2D), 'MATLAB:assertion:failed' ...
            );

            % z shape must match C & D
            Z_short = zeros([M - 1, 1]);
            Z_long = zeros([M + 1, 1]);
            testCase.verifyError( ...
                @()testCase.hfun(Z_short), 'POptUS:IncompatibleSizes' ...
            );
            testCase.verifyError( ...
                @()testCase.hfun(Z_long), 'POptUS:IncompatibleSizes' ...
            );

            % z should only by finite real vector
            for bad = [1j, 1.0 - 2.0 * 1j]
                for i = 1:length(testCase.Z)
                    z_bad = testCase.Z;
                    z_bad(i) = bad;
                    testCase.verifyError( ...
                        @()testCase.hfun(z_bad), 'POptUS:NonrealValues' ...
                    );
                end
            end

            for bad = [inf -inf nan]
                for i = 1:length(testCase.Z)
                    z_bad = testCase.Z;
                    z_bad(i) = bad;
                    testCase.verifyError( ...
                        @()testCase.hfun(z_bad), 'POptUS:NonfiniteValues' ...
                    );
                end
            end
        end

        function confirmImmutable(testCase)
            % Construct using variable declared in this scope & collect results
            C = testCase.C;
            D = testCase.D;
            [hfun] = create_censored_L1_loss_hfun(C, D);
            [hF, grads, Hash] = hfun(testCase.Z);

            % Alter same construction variable & confirm that it yields
            % different results
            C = -2.3 * C;
            [hfun_2] = create_censored_L1_loss_hfun(C, D);
            [hF_2, grads_2, Hash_2] = hfun_2(testCase.Z);
            testCase.assertNotEqual(hF_2, hF);
            testCase.assertNotEqual(grads_2, grads);
            testCase.assertNotEqual(Hash_2, Hash);

            D = -4.1 * D;
            [hfun_3] = create_censored_L1_loss_hfun(C, D);
            [hF_3, grads_3, Hash_3] = hfun_3(testCase.Z);
            testCase.assertNotEqual(hF_3, hF);
            testCase.assertNotEqual(hF_3, hF_2);
            testCase.assertNotEqual(grads_3, grads);
            testCase.assertNotEqual(grads_3, grads_2);
            testCase.assertNotEqual(Hash_3, Hash);
            testCase.assertNotEqual(Hash_3, Hash_2);

            % Confirm that changes to actual C, D arguments used to construct
            % hfun do not alter that function.  This check is motivated by
            % technical subtleties seen with Python.
            [hF_new, grads_new, Hash_new] = hfun(testCase.Z);
            testCase.assertEqual(hF_new, hF);
            testCase.assertEqual(grads_new, grads);
            testCase.assertEqual(Hash_new, Hash);
        end

    end
end

File: manifold_sampling/py/tests/TestCreateCensoredL1LossHfun.py
"""
Unit test of create_censored_L1_loss_hfun()
"""

import unittest

import numpy as np

from ibcdfo.manifold_sampling import create_censored_L1_loss_hfun


class TestCreateCensoredL1LossHfun(unittest.TestCase):
    def setUp(self):
        M = 10
        self.__SHAPE_2D = (5, 2)

        # Confirm that we have good arguments for use in testing
        self.__C = np.arange(1, M + 1).astype(float)
        self.__D = -1.1 * np.arange(1, M + 1).astype(float)

        self.__hfun = create_censored_L1_loss_hfun(self.__C, self.__D)

        # The test values were not determined by hand and are, therefore, not
        # known to be correct.  Rather, they were gathered from test output at
        # some time and are used here to catch regressions and to confirm that
        # the Python and MATLAB code are returning the same values for the same
        # problems.
        self.__Z = 2.1 * np.arange(self.__C.shape[0])
        self.__hF, self.__grads, self.__Hash = self.__hfun(self.__Z)
        self.assertEqual(156.0, self.__hF)
        expected = np.ones((M, 1))
        expected[0] = 0
        self.assertTrue(np.array_equal(self.__grads, expected))
        self.assertEqual(self.__grads.shape, (M, len(self.__Hash)))
        self.assertEqual(self.__Hash, ["2111111111"])

        # Sanity check hash result/H0
        hF_H0, grads_H0 = self.__hfun(self.__Z, self.__Hash)
        self.assertEqual(hF_H0, self.__hF)
        self.assertTrue(np.array_equal(grads_H0, self.__grads))

        # Good but different size for testing size incompatibilities
        self.__C_short = np.array([1.1, 2.2, -3.3])
        self.__D_short = np.array([3.3, -1.1, -2.2])
        M_short = len(self.__C_short)
        self.assertNotEqual(len(self.__C_short), len(self.__C))
        self.assertNotEqual(len(self.__D_short), len(self.__D))

        hfun = create_censored_L1_loss_hfun(self.__C_short, self.__D_short)

        Z_short = 2.1 * np.arange(1, self.__C_short.shape[0] + 1)
        hF, grads, Hash = hfun(Z_short)
        self.assertEqual(15.0, hF)
        expected = np.ones((M_short, 1))
        expected[0] = -1
        self.assertTrue(np.array_equal(grads, expected))
        self.assertEqual(grads.shape, (M_short, len(Hash)))
        self.assertEqual(Hash, ["311"])

        # Sanity check hash result/H0
        hF_H0, grads_H0 = hfun(Z_short, Hash)
        self.assertEqual(hF_H0, hF)
        self.assertTrue(np.array_equal(grads_H0, grads))

    def testErrors(self):
        # C & D must be NumPy arrays ...
        bad_all = [None, "bad", 1.1, [1.1], (1.1,), {1.1}, [1.1, 2.2], (1.1, 2.2), {1.1, 2.2}]
        for bad in bad_all:
            with self.assertRaises(TypeError):
                create_censored_L1_loss_hfun(bad, self.__D)
            with self.assertRaises(TypeError):
                create_censored_L1_loss_hfun(self.__C, bad)

        # but have at least 2 elements
        bad_all = [np.array([]), np.array([[]]), np.array(1.1), np.array([1.1]), np.array([[1.1]])]
        for bad in bad_all:
            with self.assertRaises(NotImplementedError):
                create_censored_L1_loss_hfun(bad, self.__D)
            with self.assertRaises(NotImplementedError):
                create_censored_L1_loss_hfun(self.__C, bad)

        # Finite reals please
        for bad_value in [np.nan, np.inf, -np.inf, 1j, 1.0 - 2.0 * 1j]:
            for i in range(len(self.__C)):
                bad = self.__C.copy()
                if isinstance(bad_value, complex):
                    bad = bad.astype(complex)
                bad[i] = bad_value
                with self.assertRaises(ValueError):
                    create_censored_L1_loss_hfun(bad, self.__D)
                with self.assertRaises(ValueError):
                    create_censored_L1_loss_hfun(self.__C, bad)

        # C & D must be *effectively* 1D ...
        C_2D = np.atleast_2d(self.__C.copy())
        D_2D = np.atleast_2d(self.__D.copy())
        self.assertEqual(2, C_2D.ndim)
        self.assertEqual(2, D_2D.ndim)
        hfun_2D = create_censored_L1_loss_hfun(C_2D, D_2D)
        hF_2D, grads_2D, Hash_2D = hfun_2D(self.__Z)
        self.assertEqual(hF_2D, self.__hF)
        self.assertTrue(np.array_equal(grads_2D, self.__grads))
        self.assertEqual(Hash_2D, self.__Hash)

        # but not actually >= 2D
        C_bad = self.__C.copy().reshape(self.__SHAPE_2D)
        D_bad = self.__D.copy().reshape(self.__SHAPE_2D)
        with self.assertRaises(ValueError):
            create_censored_L1_loss_hfun(C_bad, self.__D)
        with self.assertRaises(ValueError):
            create_censored_L1_loss_hfun(self.__C, D_bad)
        with self.assertRaises(ValueError):
            create_censored_L1_loss_hfun(C_bad, D_bad)

        # C & D must have the same effective shape
        with self.assertRaises(ValueError):
            create_censored_L1_loss_hfun(self.__C, self.__D_short)
        with self.assertRaises(ValueError):
            create_censored_L1_loss_hfun(self.__C_short, self.__D)

    def testBadZArguments(self):
        M = len(self.__C)

        # z must be NumPy array
        bad_all = [None, "bad", 1.1, (1.1,), {1.1}, [1.1, 2.2], (1.1, 2.2), {1.1, 2.2}]
        for bad in bad_all:
            with self.assertRaises(AssertionError):
                self.__hfun(bad)

        # z must be 1D
        for bad in [np.array([[]]), np.atleast_2d(self.__Z), np.zeros((M, 2))]:
            with self.assertRaises(AssertionError):
                self.__hfun(bad)

        # z must have correct length
        for bad in [np.array([]), np.ones(M - 1), np.ones(M + 1)]:
            with self.assertRaises(ValueError):
                self.__hfun(bad)

        # z must contain finite real values
        for bad_value in [np.nan, np.inf, -np.inf, 1j, 1.0 - 2.0 * 1j]:
            for i in range(len(self.__Z)):
                bad = self.__Z.copy()
                if isinstance(bad_value, complex):
                    bad = bad.astype(complex)
                bad[i] = bad_value
                with self.assertRaises(ValueError):
                    self.__hfun(bad)

    def testConfirmReadonly(self):
        # NOTE: This test is only useful for developing and manually testing the
        # code since the function under test doesn't actually alter C or D and
        # likely should never do so.
        #
        # Therefore, developers can, if so desired, temporarily inject changes
        # to C or D in the function and run this test to confirm that the
        # function is written such that it incapable of inadvertently altering
        # the C, D arrays created and managed here.
        C = self.__C.copy()
        D = self.__D.copy()
        hfun = create_censored_L1_loss_hfun(C, D)
        hfun(self.__Z)
        self.assertTrue(np.array_equal(C, self.__C))
        self.assertTrue(np.array_equal(D, self.__D))

    def testConfirmImmutable(self):
        # Construct using variable declared in this scope & collect results
        C = self.__C.copy()
        D = self.__D.copy()
        hfun = create_censored_L1_loss_hfun(C, D)
        hF, grads, Hash = hfun(self.__Z)

        # Alter same construction variables & confirm that they yield different
        # results
        C *= -2.3
        hfun_2 = create_censored_L1_loss_hfun(C, D)
        hF_2, grads_2, Hash_2 = hfun_2(self.__Z)
        self.assertNotEqual(hF_2, hF)
        self.assertFalse(np.array_equal(grads_2, grads))
        self.assertNotEqual(Hash_2, Hash)

        D *= -4.1
        hfun_3 = create_censored_L1_loss_hfun(C, D)
        hF_3, grads_3, Hash_3 = hfun_3(self.__Z)
        self.assertNotEqual(hF_3, hF)
        self.assertNotEqual(hF_3, hF_2)
        self.assertFalse(np.array_equal(grads_3, grads))
        self.assertFalse(np.array_equal(grads_3, grads_2))
        self.assertNotEqual(Hash_3, Hash)
        self.assertNotEqual(Hash_3, Hash_2)

        # Confirm that changing the construction variables didn't alter the
        # original functions
        hF_new, grads_new, Hash_new = hfun(self.__Z)
        self.assertEqual(hF_new, hF)
        self.assertTrue(np.array_equal(grads_new, grads))
        self.assertEqual(Hash_new, Hash)
